OpenRA.Mods.Common/Activities/Attack.cs

```diff
@@ -273,5 +280,18 @@ public override IEnumerable<TargetLineNode> TargetLineNodes(Actor self)
 			if (targetLineColor != null)
 				yield return new TargetLineNode(useLastVisibleTarget ? lastVisibleTarget : target, targetLineColor.Value);
 		}
+
+		bool HaveArmamentsFor(Target target)
+		{
+			var canAttack = false;
+
+			foreach (var attack in attackTraits)
+			{
+				var armaments = attack.ChooseArmamentsForTarget(target, forceAttack).ToList();
```

Comment:
`Any()` rather than fetching a whole `ToList()`?

Also while still check the remaining traits and not return at the first occurence. `if (Any()) returrn true;`.